internal/tiltfile/flags/flags.go

```diff
@@ -54,31 +98,170 @@ func setResources(thread *starlark.Thread, fn *starlark.Builtin, args starlark.T
 	return starlark.None, nil
 }
 
-var _ starkit.StatefulExtension = Extension{}
+func (e *Extension) parse(thread *starlark.Thread, fn *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	err := starkit.UnpackArgs(thread, fn.Name(), args, kwargs)
+	if err != nil {
+		return starlark.None, err
+	}
 
-func MustState(model starkit.Model) Settings {
-	state, err := GetState(model)
+	err = starkit.SetState(thread, func(settings Settings) Settings {
+		settings.flagsParsed = true
+		return settings
+	})
 	if err != nil {
-		panic(err)
+		return starlark.None, err
 	}
-	return state
+
+	st, err := starkit.GetState(thread, Settings{})
+	if err != nil {
+		return starlark.None, err
+	}
+
+	return st.(Settings).argDef.parse(e.cmdLineArgs, starkit.NewThreadWriter(thread))
 }
 
-func GetState(m starkit.Model) (Settings, error) {
-	var state Settings
-	err := m.Load(&state)
-	return state, err
+func defineStringList(thread *starlark.Thread, fn *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var name string
+	var isArgs bool
+	var usage string
+	err := starkit.UnpackArgs(thread, fn.Name(), args, kwargs,
+		"name",
+		&name,
+		"args?",
+		&isArgs,
+		"usage?",
+		&usage,
+	)
+	if err != nil {
+		return starlark.None, err
+	}
+
+	if name == "" {
+		return starlark.None, errors.New("'name' is required")
+	}
+
+	err = starkit.SetState(thread, func(settings Settings) (Settings, error) {
+		if _, ok := settings.argDef.args[name]; ok {
+			return settings, fmt.Errorf("%s defined multiple times", name)
+		}
+
+		if isArgs {
+			if settings.argDef.positionalArgName != "" {
+				return settings, fmt.Errorf("both %s and %s are defined as positional args", name, settings.argDef.positionalArgName)
+			}
+
+			settings.argDef.positionalArgName = name
+		}
+
+		settings.argDef.args[name] = argDef{
+			typ:   argTypeStringList,
+			usage: usage,
+		}
+
+		return settings, nil
+	})
+	if err != nil {
+		return starlark.None, err
+	}
+
+	return starlark.None, nil
+}
+
+type argType int
+
+const (
+	argTypeStringList argType = iota
+)
+
+type argDef struct {
+	typ   argType
+	usage string
+}
+
+type ArgsDef struct {
+	positionalArgName string
+	args              map[string]argDef
+}
+
+// Strings is a `flag.Value` for `string` arguments. (from https://github.com/sgreben/flagvar/blob/master/string.go)
+type Strings struct {
+	Values []string
+}
+
+// Set is flag.Value.Set
+func (fv *Strings) Set(v string) error {
+	fv.Values = append(fv.Values, v)
+	return nil
+}
+
+func (fv *Strings) String() string {
+	return strings.Join(fv.Values, ",")
+}
+
+func (ad ArgsDef) parse(args []string, w io.Writer) (starlark.Value, error) {
+	fs := flag.NewFlagSet("", flag.ContinueOnError)
+	fs.SetOutput(w)
+	stringListValues := make(map[string]*Strings)
```

Comment:
this function seems like it's going to get pretty gnarly if you add more types.

i wonder if there's a way it could be more modular, like if the argDef structs knew how to set up their own flags and return values?